### PR TITLE
Fix postgres migration, don't put null fields in forwards primary key

### PR DESCRIPTION
**Pull request by rustyrussell**
A bit awkward since some have already run the old migration!

Alternative to #5733

### CODE REVIEW

File: wallet/db.c [hunk 1]
```diff
@@ -940,6 +949,8 @@ static struct migration dbmigrations[] = {
     /* A reference into our own invoicerequests table, if it was made from one */
     {SQL("ALTER TABLE payments ADD COLUMN local_invreq_id BLOB DEFAULT NULL REFERENCES invoicerequests(invreq_id);"), NULL},
     /* FIXME: Remove payments local_offer_id column! */
+    /* FIXME: Remove payments local_offer_id column! */
```
Comment (vincenzopalazzo) [inline wallet/db.c:952]:
```suggestion
```  

unneeded duplication?

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/5735/commits/b8e0f76fc1970b984491268d16b618341db4d0e5

I was just now hacking around this postgress stuff, and I found that SQLite support NULL as the primary key and doing it how I was doing in https://github.com/ElementsProject/lightning/pull/5733 was messy! this is much cleaner, thanks to show me this way!

>However, to make the current version of SQLite compatible with the earlier version, SQLite allows the primary key column to contain NULL values.

**Comment by cdecker** [comment]
This is a rather heavy-handed approach to the issue imho, especially looping over thousands of rows just to create a unique fake `in_htlc_id` can take quite a while especially for larger number of HTLCs.

I see a couple of options that could work better:

 - Remove the `PRIMARY KEY` spanning both fields, and replace it with a normal `INDEX` which allows multiple entries to have the same values. This would likely not have any adverse effects, and make the `in_htlc_id` as unimportant as it truly is for bookkeeping (queries either don't use it, or it's in the `(in_htlc_id, channel_id, state)` triple in a `WHERE` clause, so not a matter of performance or consistency.
 - Use a `INSERT INTO forwards SELECT ... COALESCE(SELECT channel_htlc_id FROM channel_htlcs WHERE id = forwarded_payments.in_htlc_id, 0xE000000000 + received_time) ...`. This will not loop, but still generate unique fake IDs based on the time they were received, and we don't need
 - Not bother copying historical entries altogether

I'm not sure I like th

**Comment by denis2342** [comment]
> This is a rather heavy-handed approach to the issue imho, especially looping over thousands of rows just to create a unique fake `in_htlc_id` can take quite a while especially for larger number of HTLCs.
> 
> I see a couple of options that could work better:
> 
> * Remove the `PRIMARY KEY` spanning both fields, and replace it with a normal `INDEX` which allows multiple entries to have the same values. This would likely not have any adverse effects, and make the `in_htlc_id` as unimportant as it truly is for bookkeeping (queries either don't use it, or it's in the `(in_htlc_id, channel_id, state)` triple in a `WHERE` clause, so not a matter of performance or consistency.
> * Use a `INSERT INTO forwards SELECT ... COALESCE(SELECT channel_htlc_id FROM channel_htlcs WHERE id = forwarded_payments.in_htlc_id, -1 * received_time) ...`. This will not loop, but still generate unique fake IDs based on the time they were received, and we don't need to slowly loop over entries ourselves. Negative values would indicate a bogus ID.
> * Not bother copying historical entries altogether
> 
> I'm not sure I like the `COALESCE` trick with random bogus IDs, but it seems to work on both `sqlite3` and `postgres`, and would obviate the need for migration functions.

but it will not take long until new NULL values appear as long as there is this foreign-key constraint in table forwarded_payments:

"forwarded_payments_out_htlc_id_fkey" FOREIGN KEY (out_htlc_id) REFERENCES channel_htlcs(id) ON DELETE SET NULL

**Comment by cdecker** [comment]
Wait, I thought we delete the `forwarded_payments` table, after the migration, I might be misremembering that.

**Comment by denis2342** [comment]
> Wait, I thought we delete the `forwarded_payments` table, after the migration, I might be misremembering that.

you're right, so the "ON DELETE SET NULL" is not there anymore?

**Comment by cdecker** [comment]
Closing in favor of #5735

**Pull request closed without merging**